gpui: Add `on_button_click` / `on_any_click`

encounter opened the pull request:
`Interactivity::on_click` now accepts a `MouseButton`, indicating which button the handler is interested in.

The fluent API has two new functions: `on_button_click`, `on_any_click`. The behavior of `on_click` is preserved: meaning left-click only.

In particular, this resolves the TODO in `crates/workspace/src/pane.rs`. I'm working on a feature that adds a "middle-click to hide panel" feature, and it was blocked by the existing APIs:
- If I bound the event to `mouse_down`, it would hide the panel, but then the `mouse_up` event would get delivered to whatever was behind the panel.
- If I bound the event to `mouse_up`, middle-clicking a tab would close the tab, but then it would hide the panel when you released the mouse button.

This resolves those issues by allowing proper click handlers for any mouse button.

Release Notes:

- N/A

encounter commented:
@cla-bot check

maxdeviant reviewed (changes requested):
@encounter If we do move forward with this, I think we'll want to retain the current signature for `on_click` and have it be the left mouse button.

This is the behavior we want 90% of the time, so it adds a lot of extra overhead to have to pass in the extra parameter.

Then we can add an alternative method (e.g., `on_button_click`) that accepts a `MouseButton` and that can be used in situations where we need to register a click for a button other than left click.

encounter commented:
@maxdeviant So, given your suggestion, we'd be adding two new functions: `on_button_click`, `on_any_click`.
`on_click` would become a wrapper for `on_button_click(MouseButton::Left, ...)`

Leaving us with:
```
on_mouse_down
on_any_mouse_down
on_mouse_up
on_any_mouse_up
on_click        // calls on_button_click(MouseButton::Left, ...)
on_button_click // new
on_any_click    // new
```

Does this sound right to you?

encounter commented:
Other thoughts: maybe the signature of `Interactivity::on_click` can be updated to include the mouse button, but the fluent API (`StatefulInteractiveElement`) can have `on_click` + `on_button_click` separate?

encounter committed:
gpui: Add `on_button_click` / `on_any_click`

`Interactivity::on_click` now accepts a `MouseButton`,
indicating which button the handler is interested in.

The fluent API has two new functions: `on_button_click`,
`on_any_click`. The behavior of `on_click` is preserved:
meaning left-click only.
